integration/nwo/common/pkcs11/pkcs11.go

```diff
@@ -0,0 +1,237 @@
+	"github.com/hyperledger/fabric/bccsp"
+	"github.com/hyperledger/fabric/bccsp/pkcs11"
+	"github.com/hyperledger/fabric/bccsp/sw"
+	pkcs11lib "github.com/miekg/pkcs11"
+	"github.com/pkg/errors"
+)
+
+// GeneratePrivateKey creates a private key in the HSM and returns its corresponding public key
+func GeneratePrivateKey() (*ecdsa.PublicKey, error) {
+	lib, pin, label := FindPKCS11Lib()
+	csp, _, err := GetBCCSP(&config.BCCSP{
+		Default: "PKCS11",
+		PKCS11: &pkcs11.PKCS11Opts{
+			Security: 256,
+			Hash:     "SHA2",
+			Library:  lib,
+			Pin:      pin,
+			Label:    label,
+		},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "GeneratePrivateKey: Failed initializing PKCS11 library")
+	}
+
+	key, err := csp.KeyGen(&bccsp.ECDSAP256KeyGenOpts{Temporary: false})
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed generating ECDSA P256 key")
+	}
+
+	pub, err := key.PublicKey()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed getting public key")
+	}
+
+	raw, err := pub.Bytes()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed marshalling public key")
+	}
+
+	pk, err := DERToPublicKey(raw)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed marshalling der to public key")
+	}
+
+	return pk.(*ecdsa.PublicKey), nil
+}
+
+// GetBCCSP returns a new instance of the HSM-based BCCSP
+func GetBCCSP(conf *config.BCCSP) (bccsp.BCCSP, bccsp.KeyStore, error) {
+	if conf.PKCS11 == nil {
+		return nil, nil, errors.New("invalid config.BCCSP.PKCS11. missing configuration")
+	}
+
+	if err := CheckToken(conf.PKCS11); err != nil {
+		return nil, nil, errors.Wrap(err, "failed to check token")
+	}
+
+	p11Opts := *conf.PKCS11
+	ks := sw.NewDummyKeyStore()
+	mapper := skiMapper(p11Opts)
+	csp, err := pkcs11.New(p11Opts, ks, pkcs11.WithKeyMapper(mapper))
+	if err != nil {
+		return nil, nil, errors.WithMessagef(err, "GetBCCSP: Failed initializing PKCS11 library with config [%+v]", p11Opts)
+	}
+	return csp, ks, nil
+}
+
+func skiMapper(p11Opts config.PKCS11) func([]byte) []byte {
+	keyMap := map[string]string{}
+	for _, k := range p11Opts.KeyIDs {
+		keyMap[k.SKI] = k.ID
+	}
+
+	return func(ski []byte) []byte {
+		keyID := hex.EncodeToString(ski)
+		if id, ok := keyMap[keyID]; ok {
+			return []byte(id)
+		}
+		if p11Opts.AltID != "" {
+			return []byte(p11Opts.AltID)
+		}
+		return ski
+	}
+}
+
+// DERToPublicKey unmarshals a der to public key
+func DERToPublicKey(raw []byte) (pub interface{}, err error) {
+	if len(raw) == 0 {
+		return nil, errors.New("Invalid DER. It must be different from nil.")
+	}
+	key, err := x509.ParsePKIXPublicKey(raw)
+	return key, err
+}
+
+// ListTokens lists available tokens
+func ListTokens(opts *config.PKCS11) ([]pkcs11lib.TokenInfo, error) {
+	if opts.Library == "" {
+		return nil, fmt.Errorf("pkcs11: library path not provided")
+	}
+
+	ctx := pkcs11lib.New(opts.Library)
+	if ctx == nil {
+		return nil, fmt.Errorf("pkcs11: instantiation failed for %s", opts.Library)
+	}
+	defer ctx.Finalize()
```

Comment:
why here finalize and in row 155 it's destroy?

Reply:
good point!